### PR TITLE
Add cublet filter

### PR DESCRIPTION
**Pull request by NLGithubWP**
Skip cubelet by checking metaChunk with the following rules
1. if any field' value in `birthEvents` is not in metaChunk, then no record will birth, so skip the Cubelet. 
2. If the field's value in the cohortSelector is not in metaChunk, then all birthed records cannot be assigned to any cohort; skip the Cubelet. 
3. if the value of any fields in `valueSelector` is not in metaChunk, then even if there are birthed records and cohorts, all calculating results are empty; skip the Cubelet.

### CODE REVIEW

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/filter/Filter.java [hunk 1]
```diff
@@ -27,14 +27,14 @@ public interface Filter {
...
-    public BitSet accpet(List<String> values) throws RuntimeException;
+    public BitSet acceptStrList(List<String> values) throws RuntimeException;
     
     /**
      * Range Filter implements this function 
      * @param values the batch of input value to check
      * @return List of bool stored in BitSet
      */
-    public BitSet accept(List<Integer> values) throws RuntimeException;
+    public BitSet acceptIntList(List<Integer> values) throws RuntimeException;
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/filter/Filter.java:37]: No need to add type in function name

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/CohortProcessor.java [hunk 2]
```diff
@@ -3,60 +3,67 @@
...
 
     private String ActionTimeSchema;
     
-    private HashSet<String> projectedSchemaSet;
+    private final HashSet<String> projectedSchemaSet;
+
+    private final CohortQueryLayout layout;
 
     public CohortProcessor(CohortQueryLayout layout){
+        this.layout = layout;
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/CohortProcessor.java:60]: layout won't be used in future. Seperate the IO layer and logic layer
The memory of layout is allowed to be reclaimed.

**Review by Zrealshadow — CHANGES_REQUESTED** [review]

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/filter/SetFilter.java [hunk 3]
```diff
@@ -58,10 +61,11 @@ public Boolean accept(Integer value) throws RuntimeException {
     }
 
     @Override
-    public BitSet accpet(List<String> values) throws RuntimeException {
-        BitSet res = new BitSet(values.size());
-        for(int i = 0 ; i < values.size(); i++){
-            if (this.accept(values.get(i))) {
+    public BitSet accept(String[] values) throws RuntimeException{
+        List<String> listArr = Arrays.asList(values);
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/filter/SetFilter.java:65]: No need to transfer String[] to List<String>. Directly use String[]

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/storage/Scope.java [hunk 4]
```diff
@@ -14,7 +17,11 @@ public Scope(Integer l, Integer r){
 
     public Boolean IsInScope(Integer i){
         return i < this.right && i >= this.left;
-    }  
+    }
+
+    public Boolean IsIntersection(Scope scope){
+        return !(scope.getLeft() > right || scope.getRight() < left);
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/storage/Scope.java:23]: pay attention [left, right)
Wrong logic.
scope.getLeft() >= this.right || scope.getRight() <= this.left

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/birthSelect/EventSelection.java [hunk 5]
```diff
@@ -38,10 +38,15 @@ public boolean Accept(ProjectedTuple projectTuple) {
...
                     return false;
             } else {
+                throw new IllegalArgumentException("Only support set or range");
                 // throw execption
             }
         }
         return true;
     }
 
+    public List<Filter> getFilterList(){
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/birthSelect/EventSelection.java:48]: Just using @Getter is better

**Review by Zrealshadow — CHANGES_REQUESTED** [review]
Correct the `accept(Scope)` logic, clean the code and make sure no warnings.
Then, we can merge this PR

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/birthSelect/EventSelection.java (hunk 5 above)
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/birthSelect/EventSelection.java]: Updated

**Review by Zrealshadow — APPROVED** [review]

**Pull request merged**